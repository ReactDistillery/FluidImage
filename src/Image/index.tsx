import * as React from "react";

export interface ImageProps {
  src: string;
  alt?: string;
  title?: string;
  hasImageLoaded?: boolean;
  onLoad: Function;
}

const styles: React.CSSProperties = {
  image: {
    maxWidth: "100%",
    maxHeight: "100%",
    position: "relative",
    zIndex: 1,
    transition: "opacity 0.6s"
  }
};

const Image = ({
  onLoad,
  hasImageLoaded = false,
  title,
  alt,
  src
}: ImageProps) => (
  <img
    onLoad={event => onLoad(event)}
    src={src}
    title={title}
    style={{
      ...styles.image,
      opacity: Number(hasImageLoaded)
    }}
    alt={alt}
    draggable={false}
  />
);

export default Image;
