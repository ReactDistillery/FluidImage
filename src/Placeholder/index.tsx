import * as React from "react";

export interface PlaceholderProps {
  crop: Array<number>;
  backgroundColor?: string;
  fillContainer?: boolean;
  placeholderRef?(ref: Element | null): void;
  children?: any;
}

const styles: React.CSSProperties = {
  padding: {
    display: "inline-block",
    width: "100%",
    position: "relative",
    overflow: "hidden",
    lineHeight: 0
  }
};

const generatePlaceholderStyles = (
  crop: Array<number>,
  fillContainer = false,
  backgroundColor = "#f9f9f9"
) => ({
  ...styles.padding,
  backgroundColor: backgroundColor,
  paddingBottom: fillContainer ? 0 : `${100 * (crop[3] / crop[2])}%`,
  height: fillContainer ? "100%" : 0
});

export default class Placeholder extends React.Component<PlaceholderProps, {}> {
  render() {
    const {
      backgroundColor,
      fillContainer,
      crop,
      placeholderRef,
      children
    } = this.props;
    return (
      <span
        ref={placeholderRef}
        style={generatePlaceholderStyles(crop, fillContainer, backgroundColor)}
      >
        {children}
      </span>
    );
  }
}
