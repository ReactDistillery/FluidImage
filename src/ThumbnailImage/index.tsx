import * as React from "react";

export interface ThumbnailImageProps {
  src: string;
  alt?: string;
  title?: string;
  hasImageLoaded?: boolean;
  hideImageOnLoad?: boolean;
  onLoad: Function;
}

const styles: React.CSSProperties = {
  thumbnailImage: {
    width: "100%",
    height: "auto",
    position: "absolute",
    zIndex: 0,
    transition: "opacity 0.6s"
  }
};

const ThumbnailImage = ({
  onLoad,
  hasImageLoaded = false,
  hideImageOnLoad = false,
  title,
  alt,
  src
}: ThumbnailImageProps) => (
  <img
    onLoad={() => onLoad()}
    src={src}
    title={title}
    style={{
      ...styles.thumbnailImage,
      opacity: Number(!hideImageOnLoad && hasImageLoaded)
    }}
    alt={alt}
    draggable={false}
  />
);

export default ThumbnailImage;
