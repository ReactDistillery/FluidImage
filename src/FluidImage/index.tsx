import * as React from "react";
import Measure, {
  ContentRect,
  BoundingRect,
  MeasuredComponentProps
} from "react-measure";

import Placeholder from "../Placeholder";
import Image from "../Image";
import ThumbnailImage from "../ThumbnailImage";
import ErrorBoundary from "../ErrorBoundary";
import { dppx } from '../utils';

export interface FluidImageProps extends React.Props<any> {
  src: string;
  alt?: string;
  title?: string;
  crop: Array<number>;
  pixelRatio?: number;
  quality?: number;
  pattern?: Function;
  thumbnailSize?: number;
  backgroundColor?: string;
  thumbnailPattern?: Function;
  hideThumbnailOnLoad?: boolean;
  fillContainer?: boolean;
}

export interface Pattern {
  src: string;
  width: number;
  height: number;
  crop: Array<number>;
  pixelRatio: number;
  quality: number;
}

export interface FluidImageState {
  hasImageLoaded: boolean;
  hasThumbnailLoaded: boolean;
}

const styles: React.CSSProperties = {
  container: {
    width: "100%",
    position: "absolute"
  }
};

const setThumbnailHasLoaded = (prevState: FluidImageState) => ({
  hasThumbnailLoaded: true
});
const setImageHasLoaded = (prevState: FluidImageState) => ({
  hasImageLoaded: true
});
const setImageUnLoaded = (prevState: FluidImageState) => ({
  hasImageLoaded: false
});
const defaultPattern = ({
  src,
  width,
  crop,
  pixelRatio,
  quality,
  height
}: Pattern): string =>
  `${src}?fit=crop&w=${Math.ceil(width)}&h=${Math.ceil(
    height
  )}&dpr=${pixelRatio}&q=${quality}&rect=${crop.join(",")}`;

export default class FluidImage extends React.Component<
  FluidImageProps,
  FluidImageState
> {
  state = {
    hasImageLoaded: false,
    hasThumbnailLoaded: false
  };

  private handleImageOnLoad = () => this.setState(setImageHasLoaded);
  private handleThumbnailOnLoad = () => this.setState(setThumbnailHasLoaded);
  private handleUpdate = (bounds: BoundingRect | undefined) =>
    this.setState(setImageUnLoaded);

  render() {
    const {
      src,
      alt,
      title,
      crop,
      pixelRatio = dppx(),
      quality = 80,
      pattern = defaultPattern,
      thumbnailSize = 8,
      backgroundColor = "#f9f9f9",
      thumbnailPattern = defaultPattern,
      hideThumbnailOnLoad = false,
      fillContainer = false
    } = this.props;

    const { hasImageLoaded, hasThumbnailLoaded } = this.state;
    return (
      <ErrorBoundary
        fillContainer={fillContainer}
        crop={crop}
        backgroundColor={backgroundColor}
      >
        <Measure
          bounds
          onResize={contentRect => this.handleUpdate(contentRect.bounds)}
        >
          {({ measureRef, contentRect: { bounds } }) => (
            <Placeholder
              placeholderRef={measureRef}
              fillContainer={fillContainer}
              crop={crop}
              backgroundColor={backgroundColor}
            >
              {bounds && (
                <span style={styles.container}>
                  <ThumbnailImage
                    onLoad={this.handleThumbnailOnLoad}
                    src={thumbnailPattern({
                      src,
                      crop,
                      pixelRatio,
                      quality,
                      width: thumbnailSize,
                      height: fillContainer
                        ? bounds.height / bounds.width * thumbnailSize
                        : crop[3] / crop[2] * bounds.width * thumbnailSize
                    })}
                    hasImageLoaded={hasThumbnailLoaded}
                    hideImageOnLoad={hideThumbnailOnLoad}
                    alt={alt}
                    title={title}
                  />
                  {hasThumbnailLoaded ? (
                    <Image
                      onLoad={this.handleImageOnLoad}
                      src={pattern({
                        width: bounds!.width,
                        height: fillContainer
                          ? bounds.height
                          : crop[3] / crop[2] * bounds.width,
                        src,
                        crop,
                        pixelRatio,
                        quality
                      })}
                      title={title}
                      alt={alt}
                      hasImageLoaded={hasImageLoaded}
                    />
                  ) : null}
                </span>
              )}
            </Placeholder>
          )}
        </Measure>
      </ErrorBoundary>
    );
  }
}
