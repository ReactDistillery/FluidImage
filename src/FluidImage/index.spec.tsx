import React from "react";
import TestRenderer, { create, ReactTestRenderer } from "react-test-renderer";
import Measure from "react-measure";
import FluidImage from "./";
import ThumbnailImage from "../ThumbnailImage";
import Image from "../Image";
import Placeholder from "../Placeholder";

const minimalProps = {
  src: "testImage.jpeg",
  crop: [0, 0, 100, 100]
};

const fullProps = {
  src: "testImage.jpeg",
  crop: [0, 0, 100, 100],
  alt: "an image",
  title: "an image",
  pixelRatio: 2,
  quality: 100,
  pattern: (any: any) => "url-pattern",
  thumbnailPattern: (any: any) => "thumbnail-url-pattern",
  hideThumbnailOnLoad: true,
  fillContainer: true,
  thumbnailSize: 16,
  backgroundColor: "red"
};

const props = [minimalProps, fullProps];

const loadThumbnail = (fluidImage: ReactTestRenderer) => {
  fluidImage.root
    .findByType(ThumbnailImage)
    .findByType("img")
    .props.onLoad();
};

const loadImage = (fluidImage: ReactTestRenderer) => {
  fluidImage.root
    .findByType(Image)
    .findByType("img")
    .props.onLoad();
};

describe("FluidImage - FluidImage", () => {
  describe("render skeletons", () => {
    props.forEach((prop, index) => {
      test(`it renders with ${index} from prop list`, () => {
        const fluidImage = create(<FluidImage {...prop} />).toJSON();
        expect(fluidImage).toMatchSnapshot();
      });
    });
  });
  describe("functional behaviour", () => {
    props.forEach((prop, index) => {
      test(`it renders \`<Image />\` after hasThumbnailLoaded is true with ${index} from prop list`, () => {
        const fluidImage = create(<FluidImage {...prop} />);
        loadThumbnail(fluidImage);
        const image = fluidImage.root.findByType(Image);
        expect(image).toBeTruthy();
      });
      test(`it sets the expected state when \`<Thumbnail />\` has loaded with ${index} from prop list`, () => {
        const fluidImage = create(<FluidImage {...prop} />);
        loadThumbnail(fluidImage);
        const state = fluidImage.root.instance.state;
        expect(state).toMatchSnapshot();
      });
      test(`it sets the expected state when \`<Image />\` has loaded with ${index} from prop list`, () => {
        const fluidImage = create(<FluidImage {...prop} />);
        loadThumbnail(fluidImage);
        loadImage(fluidImage);
        const state = fluidImage.root.instance.state;
        expect(state).toMatchSnapshot();
      });
      test(`it reloads \`<Image />\` if resized with ${index} from prop list`, () => {
        const fluidImage = create(<FluidImage {...prop} />);
        loadThumbnail(fluidImage);
        loadImage(fluidImage);
        const state = fluidImage.root.instance.state;
        expect(state).toMatchSnapshot();
        fluidImage.root.findByType(Measure).props.onResize({
          bounds: {
            height: 1200,
            width: 1200
          }
        });
        const stateAfterResize = fluidImage.root.instance.state;
        expect(stateAfterResize).toMatchSnapshot();
      });
      test("it renders `Placeholder` component when compoent throws an error", () => {
        const thumbnailPattern = (any: any) => {
          throw new Error("error");
        };
        const fluidImage = create(
          <FluidImage {...prop} thumbnailPattern={thumbnailPattern} />
        );
        expect(fluidImage.root.findByType(Placeholder)).toBeTruthy();
      });
    });
  });
});
