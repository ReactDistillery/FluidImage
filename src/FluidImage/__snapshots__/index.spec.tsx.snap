// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FluidImage - FluidImage functional behaviour it reloads \`<Image />\` if resized with 0 from prop list 1`] = `
Object {
  "hasImageLoaded": true,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage functional behaviour it reloads \`<Image />\` if resized with 0 from prop list 2`] = `
Object {
  "hasImageLoaded": false,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage functional behaviour it reloads \`<Image />\` if resized with 1 from prop list 1`] = `
Object {
  "hasImageLoaded": true,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage functional behaviour it reloads \`<Image />\` if resized with 1 from prop list 2`] = `
Object {
  "hasImageLoaded": false,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage functional behaviour it sets the expected state when \`<Image />\` has loaded with 0 from prop list 1`] = `
Object {
  "hasImageLoaded": true,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage functional behaviour it sets the expected state when \`<Image />\` has loaded with 1 from prop list 1`] = `
Object {
  "hasImageLoaded": true,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage functional behaviour it sets the expected state when \`<Thumbnail />\` has loaded with 0 from prop list 1`] = `
Object {
  "hasImageLoaded": false,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage functional behaviour it sets the expected state when \`<Thumbnail />\` has loaded with 1 from prop list 1`] = `
Object {
  "hasImageLoaded": false,
  "hasThumbnailLoaded": true,
}
`;

exports[`FluidImage - FluidImage render skeletons it renders with 0 from prop list 1`] = `
<span
  style={
    Object {
      "backgroundColor": "#f9f9f9",
      "display": "inline-block",
      "height": 0,
      "lineHeight": 0,
      "overflow": "hidden",
      "paddingBottom": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <span
    style={
      Object {
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <img
      alt={undefined}
      draggable={false}
      onLoad={[Function]}
      src="testImage.jpeg?fit=crop&w=8&h=NaN&dpr=1&q=80&rect=0,0,100,100"
      style={
        Object {
          "height": "auto",
          "opacity": 0,
          "position": "absolute",
          "transition": "opacity 0.6s",
          "width": "100%",
          "zIndex": 0,
        }
      }
      title={undefined}
    />
  </span>
</span>
`;

exports[`FluidImage - FluidImage render skeletons it renders with 1 from prop list 1`] = `
<span
  style={
    Object {
      "backgroundColor": "red",
      "display": "inline-block",
      "height": "100%",
      "lineHeight": 0,
      "overflow": "hidden",
      "paddingBottom": 0,
      "position": "relative",
      "width": "100%",
    }
  }
>
  <span
    style={
      Object {
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <img
      alt="an image"
      draggable={false}
      onLoad={[Function]}
      src="thumbnail-url-pattern"
      style={
        Object {
          "height": "auto",
          "opacity": 0,
          "position": "absolute",
          "transition": "opacity 0.6s",
          "width": "100%",
          "zIndex": 0,
        }
      }
      title="an image"
    />
  </span>
</span>
`;
